File: matlab/ch07/distBPeu.m
function [x,y,T,L,V,iter] = distBPeu(opdat,mxdat)
% Calculation of multicomponent distillation using BP method (English unit)
% Set data
eos = opdat.eos; N = opdat.N; nc = opdat.nc; F = opdat.F; Tf = opdat.Tf; Pf = opdat.Pf; hF = opdat.hF;
P = opdat.P; T = opdat.T; V = opdat.V; L = opdat.L; criv = opdat.criv;
U = opdat.U; W = opdat.W; Q = opdat.Q; T0 = opdat.T0; nf = opdat.nf;
hV = opdat.hV; hL = opdat.hL; fstate = upper(opdat.fstate);
x = opdat.x; y = opdat.y; z = opdat.z; Pc = mxdat.Pc; Ant = mxdat.Ant;
% Initialization
a = zeros(1,N); b = zeros(1,N); d = zeros(1,N); V(2) = L(1) + V(1) + U(1) + W(1) - F(1);
% Feed enthalpy
for j = 1:length(nf)
[Z H phi] = phiHeu(z(:,nf(j))',Pf(nf(j)),Tf(nf(j)),fstate,eos,opdat, mxdat); hF(nf(j)) = H; % Btu/mol
end
T = T0; Told = T; criT = 10; iter = 1;
% Initialization of K(i,j): K(i,j) at T0
for j = 1:N
Pv0(:,j) = Pc(:).*exp(Ant(:,1) - Ant(:,2)./(T(j) + Ant(:,3))); K(:,j) = Pv0(:,j)./P(j); % Raoult's law
end
while criT > criv
% Calculate x(i,j)
for i = 1:nc
for j = 2:N-1
Pj(j) = V(j) - V(1) + sum(F(1:j-1) - U(1:j-1) - W(1:j-1));
Qj(j) = V(1) - (V(j) + W(j))*K(i,j) - U(j) - V(j+1) - sum(F(1:j) - U(1:j) - W(1:j));
Rj(j) = V(j+1)*K(i,j+1); Sj(j) = -F(j)*z(i,j);
end
Pj(N) = V(N) - V(1) + sum(F(1:N-1) - U(1:N-1) - W(1:N-1));
Qj(1) = V(1) - (V(1) + W(1))*K(i,1) - U(1) - V(2) - (F(1) - U(1) - W(1));
Qj(N) = V(1) - (V(N) + W(N))*K(i,N) - U(N) - sum(F(1:N) - U(1:N) - W(1:N));
Rj(1) = V(2)*K(i,2); Sj(1) = -F(1)*z(i,1); Sj(N) = -F(N)*z(i,N);
r(1) = Rj(1)/Qj(1); s(1) = Sj(1)/Qj(1);
for j = 2:N-1
r(j) = Rj(j)/(Qj(j) - r(j-1)*Pj(j)); s(j) = (Sj(j) - s(j-1)*Pj(j))/(Qj(j) - r(j-1)*Pj(j));
end
s(N) = (Sj(N) - s(N-1)*Pj(N))/(Qj(N) - r(N-1)*Pj(N)); x(i,N) = s(N);
for j = N-1:-1:1, x(i,j) = s(j) - r(j)*x(i,j+1); end
end
for j = 1:N, x(:,j) = x(:,j)/sum(x(:,j)); end
% Calculate new T(j) and y(i,j) using BP method
for j = 1:N % calculate y(i,j)
y(:,j) = K(:,j).*x(:,j); y(:,j) = y(:,j)/sum(y(:,j));
end
% Antoine eqn. (T: deg.F)
for j = 1:N % calculate T(j) (T in F)
f = @(Tv) sum(Pc'.*exp(Ant(:,1) - Ant(:,2)./(Tv +Ant(:,3))).*x(:,j)/P(j)) - 1; T(j) = fzero(f,T(j));
end
for j = 1:N
lstate = 'L';
[Z H phi] = phiHeu(x(:,j)',P(j),T(j),lstate,eos,opdat,mxdat);
phiL = phi; hL(j) = H; % Btu/lbmol
vstate = 'V'; [Z H phi] = phiHeu(y(:,j)',P(j),T(j),vstate,eos,opdat, mxdat);
phiV = phi; hV(j) = H; % Btu/lbmol
K(:,j) = phiL./phiV;
end
% Calculate loads for the condenser and reboiler
Q(1) = (L(1) + V(1) + U(1) + W(1) - F(1))*hV(2) - V(1)*hV(1) - (L(1) + U(1))*hL(1) + F(1)*hF(1);
Q(N) = sum(F.*hF - U.*hL - W.*hV) - sum(Q(1:N-1)) - V(1)*hV(1) - L(N)*hL(N);
% Calculate V(j) and L(j) sequentially
for j = 2:N
a(j) = hL(j-1) - hV(j); b(j-1) = hV(j) - hL(j-1);
d(j) = (hL(j)-hL(j-1))*sum(F(1:j-1)-U(1:j-1)-W(1:j-1)) + ...
F(j)*(hL(j)-hF(j)) + W(j)* (hV(j)-hL(j)) + Q(j);
end
a(1) = -hV(1); b(N) = -hL(N); d(1) = F(1)*(hL(1)-hF(1)) + W(1)*(hV(1)-hL(1)) + Q(1);
for j = 2:N-1
V(j+1) = (d(j) - a(j)*V(j))/b(j); L(j) = V(j+1) - V(1) + sum(F(1:j) - U(1:j) - W(1:j));
end
% Check convergence
criT = sum(abs(Told - T)); Told = T; iter = iter + 1;
end
end

File: python/ch07/distBPeu.py
import numpy as np
from scipy.optimize import fsolve
from phiHeu import phiHeu  # phiHeu.py 모듈이 있다고 가정

def distBPeu(opdat, mxdat):
    """
    BP 방법을 이용한 다성분 증류 계산 (영국 단위계)
    """
    # 데이터 설정
    eos = opdat['eos']
    N = opdat['N']
    nc = opdat['nc']
    F = opdat['F']
    Tf = opdat['Tf']
    Pf = opdat['Pf']
    P = opdat['P']
    V = opdat['V'].copy()
    L = opdat['L'].copy()
    criv = opdat['criv']
    U = opdat['U']
    W = opdat['W']
    Q = opdat['Q'].copy()
    T = opdat['T0'].copy()
    nf = opdat['nf']
    z = opdat['z']
    
    Pc = mxdat['Pc']
    Ant = mxdat['Ant']
    
    hF = np.zeros(N)
    hV = np.zeros(N)
    hL = np.zeros(N)
    x = np.zeros((nc, N))
    y = np.zeros((nc, N))
    K = np.zeros((nc, N))
    
    # 초기화: V(2) 설정
    # 파이썬 인덱스 주의 (MATLAB V(2) -> Python V[1])
    V[1] = L[0] + V[0] + U[0] + W[0] - F[0]

    # 피드 엔탈피 계산
    fstate_upper = opdat['fstate'].upper()
    for j_val in nf:
        # nf가 MATLAB 인덱스라면 1을 빼서 사용 (이미 0-index로 제공되었다고 가정)
        Z, H, phi = phiHeu(z[:, j_val], Pf[j_val], Tf[j_val], fstate_upper, eos, opdat, mxdat)
        hF[j_val] = H

    Told = T.copy()
    criT = 10.0
    iter_count = 1

    # 초기 K(i,j) 계산 (Raoult의 법칙 기반)
    for j in range(N):
        pv0 = Pc * np.exp(Ant[:, 0] - Ant[:, 1] / (T[j] + Ant[:, 2]))
        K[:, j] = pv0 / P[j]

    # 반복 계산 시작
    while criT > criv:
        # 1. 조성 x(i,j) 계산 (Thomas Algorithm/Tridiagonal Matrix Solver)
        for i in range(nc):
            Pj = np.zeros(N)
            Qj = np.zeros(N)
            Rj = np.zeros(N)
            Sj = np.zeros(N)
            
            # 중간 단 방정식 계수 설정
            for j in range(1, N-1):
                Pj[j] = V[j] - V[0] + np.sum(F[:j] - U[:j] - W[:j])
                Qj[j] = V[0] - (V[j] + W[j]) * K[i, j] - U[j] - V[j+1] - np.sum(F[:j+1] - U[:j+1] - W[:j+1])
                Rj[j] = V[j+1] * K[i, j+1]
                Sj[j] = -F[j] * z[i, j]
            
            # 첫 번째 단과 마지막 단 계수
            Qj[0] = V[0] - (V[0] + W[0]) * K[i, 0] - U[0] - V[1] - (F[0] - U[0] - W[0])
            Rj[0] = V[1] * K[i, 1]
            Sj[0] = -F[0] * z[i, 0]
            
            Pj[N-1] = V[N-1] - V[0] + np.sum(F[:N-1] - U[:N-1] - W[:N-1])
            Qj[N-1] = V[0] - (V[N-1] + W[N-1]) * K[i, N-1] - U[N-1] - np.sum(F - U - W)
            Sj[N-1] = -F[N-1] * z[i, N-1]
            
            # TDMA 풀이
            r_diag = np.zeros(N)
            s_diag = np.zeros(N)
            r_diag[0] = Rj[0] / Qj[0]
            s_diag[0] = Sj[0] / Qj[0]
            
            for j in range(1, N-1):
                denom = Qj[j] - r_diag[j-1] * Pj[j]
                r_diag[j] = Rj[j] / denom
                s_diag[j] = (Sj[j] - s_diag[j-1] * Pj[j]) / denom
            
            s_diag[N-1] = (Sj[N-1] - s_diag[N-2] * Pj[N-1]) / (Qj[N-1] - r_diag[N-2] * Pj[N-1])
            x[i, N-1] = s_diag[N-1]
            
            for j in range(N-2, -1, -1):
                x[i, j] = s_diag[j] - r_diag[j] * x[i, j+1]

        # 정규화 (Normalization)
        for j in range(N):
            x[:, j] = x[:, j] / np.sum(x[:, j])

        # 2. BP 방법을 이용한 새로운 T(j) 및 y(i,j) 계산
        for j in range(N):
            y[:, j] = K[:, j] * x[:, j]
            y[:, j] = y[:, j] / np.sum(y[:, j])

        # 버블점 온도 T 계산 (fsolve 사용)
        for j in range(N):
            def bubble_point_obj(Tv):
                pv = Pc * np.exp(Ant[:, 0] - Ant[:, 1] / (Tv + Ant[:, 2]))
                return np.sum(pv * x[:, j] / P[j]) - 1.0
            
            T[j] = fsolve(bubble_point_obj, T[j])[0]

        # K 값 업데이트
        for j in range(N):
            _, hL_j, phiL = phiHeu(x[:, j], P[j], T[j], 'L', eos, opdat, mxdat)
            hL[j] = hL_j
            _, hV_j, phiV = phiHeu(y[:, j], P[j], T[j], 'V', eos, opdat, mxdat)
            hV[j] = hV_j
            K[:, j] = phiL / phiV

        # 3. 콘덴서 및 리보일러 열부하 계산
        Q[0] = (L[0] + V[0] + U[0] + W[0] - F[0]) * hV[1] - V[0] * hV[0] - (L[0] + U[0]) * hL[0] + F[0] * hF[0]
        Q[N-1] = np.sum(F * hF - U * hL - W * hV) - np.sum(Q[:N-1]) - V[0] * hV[0] - L[N-1] * hL[N-1]

        # 4. 유량 V(j) 및 L(j) 계산
        a_vec = np.zeros(N)
        b_vec = np.zeros(N)
        d_vec = np.zeros(N)
        
        for j in range(1, N):
            a_vec[j] = hL[j-1] - hV[j]
            b_vec[j-1] = hV[j] - hL[j-1]
            d_vec[j] = (hL[j] - hL[j-1]) * np.sum(F[:j] - U[:j] - W[:j]) + \
                       F[j] * (hL[j] - hF[j]) + W[j] * (hV[j] - hL[j]) + Q[j]
        
        a_vec[0] = -hV[0]
        b_vec[N-1] = -hL[N-1]
        d_vec[0] = F[0] * (hL[0] - hF[0]) + W[0] * (hV[0] - hL[0]) + Q[0]
        
        for j in range(1, N-1):
            V[j+1] = (d_vec[j] - a_vec[j] * V[j]) / b_vec[j]
            L[j] = V[j+1] - V[0] + np.sum(F[:j+1] - U[:j+1] - W[:j+1])

        # 수렴 확인
        criT = np.sum(np.abs(Told - T))
        Told = T.copy()
        iter_count += 1

    return x, y, T, L, V, iter_count